File: frontend/src/services/api/quotes.ts
import apiClient from './client';
import { Quote, QuoteCreate, QuoteFilter, QuoteStatusUpdate } from '../../types/quote';
import { handleApiError } from '../../utils/errorHandling';
import { notification } from 'antd';

/**
 * Create a new quote
 * @param quoteData The data for the new quote
 * @returns A promise that resolves to the created quote
 */
export const createQuote = async (quoteData: QuoteCreate): Promise<Quote> => {
  try {
    const response = await apiClient.post('/quotes', quoteData);

    // Show success notification
    notification.success({
      message: 'Quote Created',
      description: `Quote has been created successfully.`,
    });

    return response.data;
  } catch (error) {
    const errorMessage = handleApiError(error, 'Failed to create quote', false);
    notification.error({
      message: 'Error Creating Quote',
      description: errorMessage,
    });
    throw error;
  }
};

/**
 * Fetch all quotes with optional filters
 * @param filters Optional filters to apply to the query
 * @returns A promise that resolves to an array of quotes
 */
export const fetchQuotes = async (filters?: QuoteFilter): Promise<Quote[]> => {
  try {
    const response = await apiClient.get('/quotes', { params: filters });
    return response.data;
  } catch (error) {
    const errorMessage = handleApiError(error, 'Failed to fetch quotes', false);
    notification.error({
      message: 'Error Fetching Quotes',
      description: errorMessage,
    });
    throw error;
  }
};

/**
 * Fetch a single quote by ID
 * @param quoteId The ID of the quote to fetch
 * @returns A promise that resolves to the quote
 */
export const fetchQuoteById = async (quoteId: number): Promise<Quote> => {
  try {
    const response = await apiClient.get(`/quotes/${quoteId}`);
    return response.data;
  } catch (error) {
    const errorMessage = handleApiError(error, `Failed to fetch quote #${quoteId}`, false);
    notification.error({
      message: 'Error Fetching Quote',
      description: errorMessage,
    });
    throw error;
  }
};

/**
 * Update the status of a quote
 * @param quoteId The ID of the quote to update
 * @param statusUpdate The status update data
 * @returns A promise that resolves to the updated quote
 */
export const updateQuoteStatus = async (
  quoteId: number,
  statusUpdate: QuoteStatusUpdate
): Promise<Quote> => {
  try {
    const response = await apiClient.patch(`/quotes/${quoteId}/status`, statusUpdate);

    // Show success notification
    notification.success({
      message: 'Quote Updated',
      description: `Quote #${quoteId} status has been updated successfully.`,
    });

    return response.data;
  } catch (error) {
    const errorMessage = handleApiError(error, `Failed to update quote #${quoteId} status`, false);
    notification.error({
      message: 'Error Updating Quote',
      description: errorMessage,
    });
    throw error;
  }
};

/**
 * Delete a quote
 * @param quoteId The ID of the quote to delete
 * @returns A promise that resolves when the quote is deleted
 */
export const deleteQuote = async (quoteId: number): Promise<void> => {
  try {
    await apiClient.delete(`/quotes/${quoteId}`);

    // Show success notification
    notification.success({
      message: 'Quote Deleted',
      description: `Quote #${quoteId} has been deleted successfully.`,
    });
  } catch (error) {
    const errorMessage = handleApiError(error, `Failed to delete quote #${quoteId}`, false);
    notification.error({
      message: 'Error Deleting Quote',
      description: errorMessage,
    });
    throw error;
  }
};

export interface QuoteGenerateRequest {
  customerId: number;
  requirements: string;
}

/**
 * Generate a quote based on customer requirements (admin only)
 * @param request The quote generation request
 * @returns A promise that resolves to the generated quote
 */
export const generateQuote = async (request: QuoteGenerateRequest): Promise<Quote> => {
  try {
    const response = await apiClient.post('/admin/quotes/generate', request);

    // Show success notification
    notification.success({
      message: 'Quote Generated',
      description: `Quote has been generated successfully for customer #${request.customerId}.`,
    });

    return response.data;
  } catch (error) {
    const errorMessage = handleApiError(error, 'Failed to generate quote', false);
    notification.error({
      message: 'Error Generating Quote',
      description: errorMessage,
    });
    throw error;
  }
};

File: warehouse_quote_app/app/api/v1/endpoints/quotes.py
"""Quote management endpoints."""
from fastapi import APIRouter, Depends, HTTPException, status, Query, Path
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime, timezone
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession

from warehouse_quote_app.app.core.auth import get_current_user
from warehouse_quote_app.app.api import deps
from warehouse_quote_app.app.models.user import User
from warehouse_quote_app.app.schemas.quote import (
    QuoteCreate,
    QuoteResponse,
    QuoteStatusUpdate,
    QuoteListResponse,
    QuoteFilter,
    QuoteNegotiationRequest,
    QuoteNegotiationResponse
)
from warehouse_quote_app.app.services.quote_lifecycle import QuoteLifecycleService
from warehouse_quote_app.app.core.logging import get_logger
from warehouse_quote_app.app.api.v1.middleware import rate_limit, cache_response, RATE_LIMIT_STANDARD

router = APIRouter()
logger = get_logger("quotes")

@router.post(
    "/quotes",
    response_model=QuoteResponse,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(rate_limit(RATE_LIMIT_STANDARD))],
    responses={
        201: {"description": "Quote created successfully"},
        400: {"description": "Invalid input"},
        401: {"description": "Unauthorized"},
        403: {"description": "Forbidden"},
        500: {"description": "Internal server error"}
    }
)
async def create_quote(
    quote_data: QuoteCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(deps.get_db)
):
    """Create a new quote."""
    try:
        quote_service = QuoteLifecycleService(db)
        return await quote_service.create_quote(quote_data, current_user.id)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        logger.error(f"Error creating quote: {str(e)}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Error creating quote")

@router.get(
    "/quotes/{quote_id}",
    response_model=QuoteResponse,
    dependencies=[Depends(rate_limit(RATE_LIMIT_STANDARD))],
    responses={
        200: {"description": "Quote retrieved successfully"},
        401: {"description": "Unauthorized"},
        403: {"description": "Forbidden"},
        404: {"description": "Quote not found"},
        500: {"description": "Internal server error"}
    }
)
async def get_quote(
    quote_id: int = Path(..., title="Quote ID", ge=1),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(deps.get_db)
):
    """Get a quote by ID."""
    try:
        quote_service = QuoteLifecycleService(db)
        return await quote_service.get_quote(quote_id)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except Exception as e:
        logger.error(f"Error retrieving quote: {str(e)}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Error retrieving quote")

@router.get(
    "/quotes",
    response_model=QuoteListResponse,
    dependencies=[Depends(rate_limit(RATE_LIMIT_STANDARD))],
    responses={
        200: {"description": "Quotes retrieved successfully"},
        401: {"description": "Unauthorized"},
        403: {"description": "Forbidden"},
        500: {"description": "Internal server error"}
    }
)
async def list_quotes(
    skip: int = Query(0, title="Skip", description="Number of quotes to skip", ge=0),
    limit: int = Query(100, title="Limit", description="Maximum number of quotes to return", ge=1, le=100),
    status: Optional[str] = Query(None, title="Status", description="Filter by quote status"),
    start_date: Optional[datetime] = Query(None, title="Start Date", description="Filter by start date"),
    end_date: Optional[datetime] = Query(None, title="End Date", description="Filter by end date"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(deps.get_db)
):
    """List all quotes with filtering."""
    try:
        quote_service = QuoteLifecycleService(db)
        
        # Create filter object if any filter parameters are provided
        filter_params = None
        if status or start_date or end_date:
            filter_params = QuoteFilter(
                status=status,
                start_date=start_date,
                end_date=end_date
            )
            
        return await quote_service.list_quotes(skip, limit, filter_params)
    except Exception as e:
        logger.error(f"Error listing quotes: {str(e)}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Error listing quotes")

@router.patch(
    "/quotes/{quote_id}/status",
    response_model=QuoteResponse,
    dependencies=[Depends(rate_limit(RATE_LIMIT_STANDARD))],
    responses={
        200: {"description": "Quote status updated successfully"},
        400: {"description": "Invalid input"},
        401: {"description": "Unauthorized"},
        403: {"description": "Forbidden"},
        404: {"description": "Quote not found"},
        500: {"description": "Internal server error"}
    }
)
async def update_quote_status(
    status_update: QuoteStatusUpdate,
    quote_id: int = Path(..., title="Quote ID", ge=1),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(deps.get_db)
):
    """Update quote status."""
    try:
        quote_service = QuoteLifecycleService(db)
        return await quote_service.update_quote_status(quote_id, status_update)
    except ValueError as e:
        if "not found" in str(e).lower():
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
        else:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        logger.error(f"Error updating quote status: {str(e)}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Error updating quote status")

@router.post(
    "/quotes/{quote_id}/negotiate",
    response_model=QuoteNegotiationResponse,
    status_code=status.HTTP_201_CREATED,
    dependencies=[Depends(rate_limit(RATE_LIMIT_STANDARD))],
    responses={
        201: {"description": "Negotiation request submitted successfully"},
        400: {"description": "Invalid input"},
        401: {"description": "Unauthorized"},
        403: {"description": "Forbidden"},
        404: {"description": "Quote not found"},
        500: {"description": "Internal server error"}
    }
)
async def negotiate_quote(
    negotiation_request: QuoteNegotiationRequest,
    quote_id: int = Path(..., title="Quote ID", ge=1),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(deps.get_db)
):
    """Submit a negotiation request for a quote."""
    try:
        quote_service = QuoteLifecycleService(db)
        return await quote_service.negotiate_quote(
            quote_id, 
            current_user.id,
            negotiation_request
        )
    except ValueError as e:
        if "not found" in str(e).lower():
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
        else:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        logger.error(f"Error negotiating quote: {str(e)}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, 
            detail="Error submitting quote negotiation request"
        )

@router.delete(
    "/quotes/{quote_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(rate_limit(RATE_LIMIT_STANDARD))],
    responses={
        204: {"description": "Quote deleted successfully"},
        401: {"description": "Unauthorized"},
        403: {"description": "Forbidden"},
        404: {"description": "Quote not found"},
        500: {"description": "Internal server error"}
    }
)
async def delete_quote(
    quote_id: int = Path(..., title="Quote ID", ge=1),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(deps.get_db)
):
    """Delete a quote."""
    try:
        quote_service = QuoteLifecycleService(db)
        result = await quote_service.delete_quote(quote_id)
        if not result:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Quote with ID {quote_id} not found")
        return JSONResponse(status_code=status.HTTP_204_NO_CONTENT, content=None)
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting quote: {str(e)}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Error deleting quote")
